### ISSUE TITLE
LR Finder doesn't restore original model weights?

### ISSUE THREAD

**Issue by AmarSaini** [davidtvs/pytorch-lr-finder#11]
Hey! I love this repo, thanks for making it 💯 

Everything works well except for one thing, after some digging around/experimenting, here's what I've found:

Below are some figures for the training loss and training accuracy (on MNIST, using a resnet18).

Problem:

1) Using LRFinder on a model, and then training with it afterwards appears to hurt the models learning (see pink curve below).

Solution:

2) Using LRFinder on a model, and _manually_ restoring the weights, appears to train the model optimally. (see green curve below).
3) Using LRFinder on a clone of the model, and then using the original model for training, appears to train the model optimally. (see green curve below).

Regarding the figure/graphs below, both models used the same hyperparameters.

An in-code example of option 1) would be similar to what was given in the README.md:
```
from torch_lr_finder import LRFinder

model = ...
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-7, weight_decay=1e-2)
lr_finder = LRFinder(model, optimizer, criterion, device="cuda")
lr_finder.range_test(trainloader, end_lr=100, num_iter=100)
lr_finder.plot()

// Then use "model" for training

```
An in-code example of option 3) would be:
```
from torch_lr_finder import LRFinder

model = ...
temp_model = *create model with same architecture*
// copy weights over
temp_model.load_state_dict(model.state_dict)
criterion = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=1e-7, weight_decay=1e-2)
// use temp model in lr_finder
lr_finder = LRFinder(temp_model, optimizer, criterion, device="cuda")
lr_finder.range_test(trainloader, end_lr=100, num_iter=100)
lr_finder.plot()
```

![image](https://user-images.githubusercontent.com/25353789/71288840-6e057b00-2320-11ea-83d7-134b534528e0.png)

**Comment by NaleRaphael** [comment]
Hi @AmarSaini , thanks for these detailed report!

However, it seems like `lr_finder.reset()` wasn't called after `lr_finder.range_test()` finished. So that model and optimizer were not restored to their initial state. And that's probably the reason why the model performs bad after it.

Could you try to call `lr_finder.reset()` after the process of learning rate finding is finished, and check whether the model works normally or not?

If it still not work, please feel free to let us know!

**Comment by AmarSaini** [comment]
Wow, totally didn't see that haha. I did go through some of the source code and saw some code regarding restoring original model/opt parameters. I thought this was automatic (as in FastAI), my fault for not reading the complete README.md. Will try it again later tonight and follow up!

**Comment by AmarSaini** [comment]
Also, forgot to ask. In fastai we can run lr_find after multiple training sessions.

With this API, we have to supply the optimizer ourselves. Just want to confirm that this means we should make a new (temp) optimizer, and copy the state from the current optimizer (used in training) over to the the new (temp) optimizer to use in LR_Find. (So lr_find is testing the loss with respect to different learning rates with the same optimizer state being used in training). This would be the correct thing to do right?

Or do you recommend using a fresh optimizer for LR_Find, even if you are planning to continue training?

Thanks :)

**Comment by NaleRaphael** [comment]
That's totally fine!
In my opinion, providing an API like `lr_finder.reset()` makes user able to do some post-processing (e.g. dump the weight for experimental test or something else) before the model and optimizer is restored.
As for the implementation in fastai, it's actually a great design for user to build a training pipeline.

For the question about using `LRFinder` in multiple training sessions, it doesn't need to make a new optimizer. But we have to create a new `LRFinder` in each new training session due to the implementation shown as below:
https://github.com/davidtvs/pytorch-lr-finder/blob/4de54de6b799680d87ac3b7dec2752f20f8c6306/torch_lr_finder/lr_finder.py#L52-L57

Initial state of model and optimizer **is only able to be cached when a `LRFinder` is instantiated**. That's why we have to create a new `LRFinder` each time if we need to do learning rate finding in multiple training sessions (e.g. find learning rate after every 4 epochs of training).

In a nutshell, we should do like this in multiple training sessions:
```
# --- session 01 ---
lr_finder = LRFinder(...)
lr_finder.range_test(...)
lr_finder.reset()  # reset model and optimizer to their own initial state

# update learning rate of optimizer
best_lr = find_best_lr(lr_finder.history)  # a custom method to find best learning rate from history
for param_group in optimizer.param_groups:
    param_group['lr'] = best_lr

# start training model...


# --- session 02 ---
lr_finder = LRFinder(...)
lr_finder.range_test(...)
lr_finder.reset()

# update learning rate of optimizer
best_lr = find_best_lr(lr_finder.history)
for param_group in optimizer.param_groups:
    param_group['lr'] = best_lr

# start training model...
```

Hope these information help! :)

**Comment by AmarSaini** [comment]
Thanks for the detailed response :)

Regarding the multiple training sessions to continue training, is there functionality in this API to reuse an optimizer?

(Before running session 2's LR Finder: I have to reset the learning rate to 1e-7 while also maintaining the optimizer state).

E.g.
```
for param_group in optimizer.param_groups:
    param_group['lr'] = 1e-7
    // Also reset other optimizer parameters to their defaults, e.g. momentum

**Comment by NaleRaphael** [comment]
Yes, you are right! We have to reset learning rate of the optimizer before running `lr_finder.range_test()` in the second session. Otherwise, `lr_finder.range_test()` will start the scanning from the latest learning rate used in previous session. (like the implementation in [`fastai.callbacks.lr_finder`](https://github.com/fastai/fastai/blob/330cbd34e2d597528784f8603112ba94b721c289/fastai/callbacks/lr_finder.py#L21), learning rate will be reset to the initial value)

But I'm sorry about that there is currently no API to reset the hyperparameters of an optimizer automatically. Because that there are kinds of different implementation of optimizer, and we cannot determine what hyperparameters have to be reset. So this operation should be done by users.

However, it seems to be an interesting challenge to do so. As far as I know, fastai has implemented a wrapper for optimizer (see also: [OptimWrapper](https://github.com/fastai/fastai/blob/330cbd34e2d597528784f8603112ba94b721c289/fastai/callback.py#L9-L157)).
I'll keep this in mind and maybe send a new PR for this if there is further breakthrough of design in `torch.optim`.

I'm glad to discuss with you! :)

**Comment by AmarSaini** [comment]
Thanks for all the info :)

And no worries, it's not too bad to reset the optimizer. As for it being trouble since different optimizers have different hyperparameters, here is how I would get around it:

```
optim_for_lr_find = torch.optim.*any optimizer with any settings*

optim_for_training = *some optim being used for training*

temp_dict = {}
temp_dict['state'] = optim_for_training.state_dict['state']
temp_dict['param_groups'] = optim_for_lr_find['param_groups']

optim_for_lr_find.load_state_dict(temp_dict)

// Run your LR Find with optim_for_lr_find
```
After plotting lr find, we can find a new optimal lr and put it in our current optimizer being used for training, without changing it's state:

```
optim_for_new_lr = torch.optim.*optimizer with optimal lr*

optim_for_training = *some optim being used for training*

temp_dict = {}
temp_dict['state'] = optim_for_training.state_dict['state']
temp_dict['param_groups'] = optim_for_new_lr['param_groups']

optim_for_training.load_state_dict(temp_dict)
```

**Comment by NaleRaphael** [comment]
That's a good idea!

And I've run some simple snippet to check the content of `optimizer.param_groups`:
```python
# for `optim.SGD`
>>> optimizer.param_groups[0].keys()
dict_keys(['lr', 'momentum', 'dampening', 'weight_decay', 'nesterov', 'params'])

# for `optim.Adam`
>>> optimizer.param_groups[0].keys()
dict_keys(['params', 'lr', 'betas', 'eps', 'weight_decay', 'amsgrad', 'initial_lr'])
```

Since all items in `optimizer.param_groups[0]` (except `optimizer.param_groups[0]['params']`) are the hyperparameters of an optimizer, it seems that it's possible to make a generalized solution to reset them.

But I have to take more time to dive into the source code of pytorch to figure it out.
Thanks for your help! ;)

**Comment by AmarSaini** [comment]
Oh nice!

Let me know if you're looking for any help, I'd be more than happy to contribute on any part of this lovely repo! :)

**Comment by NaleRaphael** [comment]
Sure! I'll update this thread once there is any breakthrough!

**Comment by davidtvs** [comment]
Is there a need to reset other parameters other than the learning rate?

Seems to me that you would always want to keep everything else as-is and just lower the learning rate. If that's the case then we can add an `initial_lr` parameter to `range_test`. See the code example below changed from what @NaleRaphael posted before:

```python
# --- session 01 ---
optimizer = torch.optim... # Some optimizer
lr_finder = LRFinder(model, optimizer, ...)
lr_finder.range_test(initial_lr=None, ...) # Set to None defaults to the lr in optimizer
lr_finder.reset()  # reset model and optimizer to their own initial state

# update learning rate of optimizer
best_lr = find_best_lr(lr_finder.history)  # a custom method to find best learning rate from history
for param_group in optimizer.param_groups:
    param_group['lr'] = best_lr

# start training model...


# --- session 02 ---
lr_finder = LRFinder(model, optimizer, ...) # optimizer was used for training so it carries the state from training
lr_finder.range_test(initial_lr=1e-7, ...) # change the initial learning rate to something else
lr_finder.reset() # resets the optimizer to the state and parameters after training

# update learning rate of optimizer
best_lr = find_best_lr(lr_finder.history)
for param_group in optimizer.param_groups:
    param_group['lr'] = best_lr
```

**Comment by AmarSaini** [comment]
Hi! I'm not too sure how important it is if we preserve the other hyper-parameters in the optimizer. Currently, I'm using the OneCycle LR Scheduler with the Adam Optimizer.

Before training this is what my optimizer looks like for using LR Finder (no lr scheduler), using `lr=1e-7`:
<img src="https://user-images.githubusercontent.com/25353789/71327881-f885e000-24c2-11ea-91b8-0b405eaf06f0.png" width="40%">

After find a good learning rate, `lr=1e-3` training this is what my optimizer looks after wrapped with the OneCycle LR Scheduler:
<img src="https://user-images.githubusercontent.com/25353789/71327632-c8890d80-24bf-11ea-84df-0fbadb455d1f.png" width="40%">

After training session 1, this is what my optimizer looks like:
<img src="https://user-images.githubusercontent.com/25353789/71327769-72b56500-24c1-11ea-8699-70c58b949abd.png" width="40%">

Before training session 2, let's say we want to find another learning rate.
Will a wrapped LR Scheduler effect the LR Finder?
Simply reseting the `lr` back to `lr=1e-7` would yield this for the optimizer:
<img src="https://user-images.githubusercontent.com/25353789/71327943-3df6dd00-24c4-11ea-8b9c-8ddb21351d4f.png" width="40%">

Running Find LR with this resetted `lr=1e-7` optimizer wrapped with a LR Scheduler worked fine:
<img src="https://user-images.githubusercontent.com/25353789/71328058-aabea700-24c5-11ea-8a42-4e422f72729e.png" width="40%">

Then we can set the `lr` to something like `lr=1e-3` (starting LR for OneCycle LR Scheduler uses the `lr` of the optimizer, my `max_lr` would be `lr=1e-2`
<img src="https://user-images.githubusercontent.com/25353789/71328064-be6a0d80-24c5-11ea-9bdc-4019f413538c.png" width="60%">

Then finally, re-wrap with the OneCycle LR Scheduler to get:
<img src="https://user-images.githubusercontent.com/25353789/71328073-d5a8fb00-24c5-11ea-85a4-1006b295782d.png" width="40%">

So it looks like this would work nicely. My current solution was just to make a temp optimizer to get default param groups, and manually copy the state over to it.

**Comment by NaleRaphael** [comment]
@davidtvs :
> Is there a need to reset other parameters other than the learning rate?

If there is an optimizer of which the hyperparameters will change according to iteration, then we might need to reset it. But I am not sure whether there is a kind of this optimizer up to now.
 
> Seems to me that you would always want to keep everything else as-is and just lower the learning rate. If that's the case then we can add an `initial_lr` parameter to `range_test`.

This sounds great!

@AmarSaini :
> Will a wrapped LR **Scheduler** (optimizer?) effect the LR Finder?

As far as I know, `lr` of an optimizer which is wrapped by `torch.optim.lr_scheduler.*` will change in only 2 stage:
1. when `lr_scheduler.step()` is called (see also [torch.optim.lr_scheduler#L88-89](https://github.com/pytorch/pytorch/blob/v1.2.0/torch/optim/lr_scheduler.py#L88-L89))
2. when a `lr_scheduler` is just initialized (see also [torch.optim.lr_scheduler#L43](https://github.com/pytorch/pytorch/blob/v1.2.0/torch/optim/lr_scheduler.py#L43))

With the point 2 mentioned above, we should take care that the beginning `lr` used by `lr_finder` will be affected if an optimizer has been wrapped by a `lr_scheduler` before running `lr_finder.range_test()`.
For this reason, I would suggest performing `lr_finder.range_test()` before wrapping a optimizer by any `lr_scheduler` to avoid unexpected result.

**Comment by AmarSaini** [comment]
@davidtvs 
>For this reason, I would suggest performing lr_finder.range_test() before wrapping a optimizer by any lr_scheduler to avoid unexpected result.

Would this mean the end-user should be responsible for un-wrapping an optimizer before calling `lr_finder.range_test()` a second time before another training session?

**Comment by davidtvs** [comment]
@AmarSaini Ya, currently, the end-user should take care of it.
We could change that in the future though; check if the optimizer has a scheduler wrapped around it and do something like what you described. I can look into it, or review a PR for this

**Comment by AmarSaini** [comment]
Ah I see!

Not sure if it's worth the trouble to implement this :P
Is there any work out there that you know of that shows the importance of re-using the same optimizer for continued training?

I made a post on the PyTorch forums, but no replies yet: https://discuss.pytorch.org/t/importance-of-optimizers-when-continuing-training/64788 

Happy Holidays by the way!

**Comment by NaleRaphael** [comment]
@AmarSaini :
> Would this mean the end-user should be responsible for un-wrapping an optimizer before calling `lr_finder.range_test()` a second time before another training session?

I agree with @davidtvs though it may be inconvenience to un-wrap an optimizer manually currently. But it seems good to make this change to avoid unexpected result for most of the users.

If we are going to work on this, it seems like we can check whether an optimizer has been wrapped by a lr scheduler according to the key `"initial_lr"` in `optimizer.param_groups[0]`. (see also [`torch.optim.lr_scheduler._LRScheduler.__init__`; v0.4.1](https://github.com/pytorch/pytorch/blob/v0.4.1/torch/optim/lr_scheduler.py#L15-L22))

> Is there any work out there that you know of that shows the importance of re-using the same optimizer for continued training?

I'm not sure about this, too. Based on my past experience, I would use the same optimizer to continue training even after the training process is paused. But I would also reload the dumped state from checkpoint into model and optimizer before continuing.

I'll keep studying on this weekend. ;)

**Comment by NaleRaphael** [comment]
Sorry for the late response.

> Is there any work out there that you know of that shows the importance of re-using the same optimizer for continued training?

I've wrote a script for this [(open in colab)](https://colab.research.google.com/drive/1mWrVrOmcAr8YaCzVD5t_P4igT0t4lSIt), but got no significant result. There is no too much difference between the validation loss and accuracy in both cases (using the original optimizer or a new one for fine-tuning a model).
But it seems fine after re-inspecting the implementation of optimizer, since there is no iteration-chaning attribute in it. Although we tried to fine tune a model with a new optimizer, we still initialize it with the weight of trained model. So that there is no negative effect if we doing so.

**Comment by AmarSaini** [comment]
Hey no worries!

Very interesting notebook and thanks for sharing! I also ran into my loss increasing when continuing training, very weird. Came across [this post](https://ai.stackexchange.com/questions/8063/loss-jumps-abruptly-when-i-decay-the-learning-rate-with-adam-optimizer-in-pytorc) if you're interested in that too.

But yea, it looks like maintaining optimizer state (or ditching it and making a new optimizer) doesn't seem to affect things too much.

I also plan on maybe running some experiments, probably on a bigger model with a harder dataset (I worry that we don't see a difference because the task is too easy :P). Will follow up on this and confirm if I got the same results!

Also, I love the object oriented layout of your notebook, it really follows the FastAI style (putting everything in a Learner object). Out of curiousity, is there a template online for this, or did you make this yourself over time?

**Comment by NaleRaphael** [comment]
Thanks for sharing the post! That's truly a weird phenomenon, like the answerer Dennis Soemers said in it.

I feel the same that the problem may be over-simplified by using only MNIST dataset or such a small convolutional network. Besides, the strategy used for fine-tuning a model might be related, e.g. freeze most of the layers from the top of model at the beginning of fine-tuning, then unfreeze them gradually after a few epochs/iterations. I could do some more experiments for this.

And the `Learner` object is made by myself after doing some research on the framework developed by FastAI. I really like the design they made for easily building a training pipeline, so I made a simplified one for this experiment. Though it's just a toy-like implementation, I'm really appreciate you like it!

**Comment by AmarSaini** [comment]
So I ran some experiments too, check out my project page: [Optimizer Benchmarks](https://amarsaini.github.io/Optimizer-Benchmarks/)

Main conclusion: They both achieve similar performance. (1: Making new optimizers. Or 2: Preserving state and re-using the same optimizer).

**Comment by davidtvs** [comment]
I made a PR (#14) to add allow changing the starting learning rate when calling `range_test()`. Feedback is welcome, either here or in the PR

**Comment by davidtvs** [comment]
Closing this, for now, great discussion. Feel free to reopen if your use-case requires other parameters besides the learning rate to be reset
